File: 9.Caching, the secret behind it all/code/python/104_session_management_with_redis_python_1.py
import json
import functools
import redis
from fastapi import FastAPI

app = FastAPI()

# Connect to Redis
r = redis.Redis(host="localhost", port=6379, decode_responses=True)

def cache_result(ttl: int = 3600):
    """
    Decorator that implements lazy (cache-aside) caching.
    ttl: time-to-live in seconds (default 1 hour)
    """
    def decorator(func):
        @functools.wraps(func)
        async def wrapper(*args, **kwargs):
            # Build cache key from function name + args
            cache_key = f"{func.__name__}:{args}:{kwargs}"

            # Step 1: Check cache (Cache Hit)
            cached = r.get(cache_key)
            if cached:
                print(f"[CACHE HIT] {cache_key}")
                return json.loads(cached)

            # Step 2: Cache Miss ,  execute the actual function
            print(f"[CACHE MISS] calling {func.__name__}...")
            result = await func(*args, **kwargs)

            # Step 3: Store in Redis with TTL
            r.setex(cache_key, ttl, json.dumps(result))

            return result
        return wrapper
    return decorator


# Usage: apply cache decorator to any route handler
@app.get("/products/{product_id}")
@cache_result(ttl=3600)  # cache for 1 hour
async def get_product(product_id: str):
    # Expensive DB query ,  only runs on cache miss
    product = await fetch_from_db(product_id)
    return product


# TTL-based API Response Caching (e.g. weather)
@app.get("/weather/{city}")
async def get_weather(city: str):
    cache_key = f"weather:{city}"

    # Check cache (TTL = 1 hour, weather doesn't change every minute)
    cached = r.get(cache_key)
    if cached:
        return {"source": "cache", "data": json.loads(cached)}

    # Miss: call external weather API (costs money / rate limited)
    weather_data = await call_weather_api(city)

    # Store for 1 hour
    r.setex(cache_key, 3600, json.dumps(weather_data))

    return {"source": "api", "data": weather_data}

File: 9.Caching, the secret behind it all/code/go/104_session_management_with_redis_python_1.go
package main

import (
    "context"
    "encoding/json"
    "fmt"
    "time"

    "github.com/redis/go-redis/v9"
)

type Product struct {
    ID    string `json:"id"`
    Name  string `json:"name"`
    Price float64 `json:"price"`
}

var rdb = redis.NewClient(&redis.Options{
    Addr: "localhost:6379",
})

// GetProduct implements Cache-Aside (Lazy) caching.
// 1. Check Redis first
// 2. On miss, fetch from DB, store in cache, return
func GetProduct(ctx context.Context, productID string) (*Product, error) {
    cacheKey := "product:" + productID

    // Step 1: Try cache first (Cache Hit path)
    cached, err := rdb.Get(ctx, cacheKey).Result()
    if err == nil {
        var product Product
        json.Unmarshal([]byte(cached), &product)
        fmt.Println("[CACHE HIT]", productID)
        return &product, nil
    }

    // Step 2: Cache Miss ,  fetch from database (expensive operation)
    fmt.Println("[CACHE MISS] fetching from DB...", productID)
    product, err := fetchFromDatabase(productID) // simulate DB call
    if err != nil {
        return nil, err
    }

    // Step 3: Store in cache with a 1-hour TTL
    data, _ := json.Marshal(product)
    rdb.Set(ctx, cacheKey, data, 1*time.Hour)

    return product, nil
}

// Write-Through: update DB and cache simultaneously
func UpdateProduct(ctx context.Context, product *Product) error {
    // Step 1: Update in database
    if err := updateInDatabase(product); err != nil {
        return err
    }

    // Step 2: Write-through ,  update cache immediately
    cacheKey := "product:" + product.ID
    data, _ := json.Marshal(product)
    rdb.Set(ctx, cacheKey, data, 1*time.Hour)

    fmt.Println("[WRITE-THROUGH] DB + cache updated for", product.ID)
    return nil
}
